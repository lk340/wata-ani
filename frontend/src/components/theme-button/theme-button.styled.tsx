import styled from "styled-components";
import { animated } from "react-spring";

import * as Constants from "@/utils/style/constants";
import * as Snippets from "@/utils/style/snippets";

import lightMode from "@/icons/navbar/light-mode.svg";
import darkMode from "@/icons/navbar/dark-mode.svg";

export const ThemeButton = styled("div").attrs(() => ({
	"data-testid": "theme button component",
}))`
	${Snippets.flexRowCenter()};
	position: relative;
	margin: 0px ${Constants.size.components.navbar.spacer}px;
	cursor: pointer;
`;

export const ThemeButtonLightModeIcon = styled(animated.img).attrs(() => ({
	src: lightMode,
	alt: "navbar component light theme button",
}))`
	${Snippets.square(`${Constants.size.components.navbar.icon}px`)};
	position: absolute;
`;

export const ThemeButtonDarkModeIcon = styled("img").attrs(() => ({
	src: darkMode,
	alt: "navbar component dark theme button",
}))`
	${Snippets.square(`${Constants.size.components.navbar.icon}px`)};
`;
