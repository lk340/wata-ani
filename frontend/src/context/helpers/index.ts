import { createUseContext } from "./create-use-context";

export { createUseContext };
