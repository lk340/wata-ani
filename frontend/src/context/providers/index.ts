import { Providers } from "./_providers";
export { Providers };
