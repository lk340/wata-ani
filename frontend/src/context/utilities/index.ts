import * as Theme from "./_theme";
import * as Window from "./_window";
import * as WindowScroll from "./_scroll";
import * as Mouse from "./_mouse";
import * as Location from "./_location";

export { Theme, Window, WindowScroll, Mouse, Location };
