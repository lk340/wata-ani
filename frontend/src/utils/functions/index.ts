// =============================== //
// ↓↓↓ Data Structures Utility ↓↓↓ //
// =============================== //

export function removeArrayElement<ArrayElement>(
	array: ArrayElement[],
	index: number,
): ArrayElement[] {
	array.splice(index, 1);
	return array;
}

export function capitalizeFirstLetter(wordOrPhrase: string): string {
	const firstLetter = wordOrPhrase[0].toUpperCase();
	const restOfWordOrPhrase = wordOrPhrase.slice(1);
	return firstLetter + restOfWordOrPhrase;
}
