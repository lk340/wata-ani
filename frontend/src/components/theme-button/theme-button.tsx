import * as React from "react";

import * as Context from "@/context";
import * as Animations from "@/utils/style/animations";

import * as Styled from "./theme-button.styled";

export const ThemeButton = () => {
	const { theme } = Context.Theme.useThemeContext();

	const animateLightModeIcon = Animations.opacity(theme.state.mode === "dark");

	return (
		<Styled.ThemeButton onClick={theme.setters.toggleMode}>
			<Styled.ThemeButtonLightModeIcon style={animateLightModeIcon} />
			<Styled.ThemeButtonDarkModeIcon />
		</Styled.ThemeButton>
	);
};
