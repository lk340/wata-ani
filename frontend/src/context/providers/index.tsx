import * as React from "react";

import * as Context from "@/context";

export const Providers: React.FC<{ children: React.ReactNode }> = ({ children }) => {
	return (
		<>
			<Context.Theme.Provider>
				<Context.Windows.Provider>
					<Context.Scroll.Provider>
						<Context.Mouse.Provider>
							<Context.Location.Provider>
								<Context.UserAgent.Provider>
									<Context.AuthForm.Provider>
										<Context.Navbar.Provider>
											<Context.Pagination.Provider>
												<Context.NavbarOptionsCreate.Provider>
													<Context.NavbarOptionsSettings.Provider>
														{children}
													</Context.NavbarOptionsSettings.Provider>
												</Context.NavbarOptionsCreate.Provider>
											</Context.Pagination.Provider>
										</Context.Navbar.Provider>
									</Context.AuthForm.Provider>
								</Context.UserAgent.Provider>
							</Context.Location.Provider>
						</Context.Mouse.Provider>
					</Context.Scroll.Provider>
				</Context.Windows.Provider>
			</Context.Theme.Provider>
		</>
	);
};

// const providers = [
// 	Context.Theme.Provider,
// 	Context.Window.Provider,
// 	Context.Scroll.Provider,
// 	Context.Mouse.Provider,
// 	Context.Location.Provider,
// 	Context.Auth.Provider,
// 	Context.UserAgent.Provider,
// 	Context.Filter.Provider,
// 	Context.Sort.Provider,
// 	Context.Demo.Provider,
// 	Context.Home.Provider,
// ];

// function generateProvider(providers: React.ReactNode[]) {
// 	if (providers.length === 1) return providers[0];

// 	const lastIndex = providers.length - 1;

// }

// const Children: React.FC<{ children: React.ReactNode }> = ({ children }) => {
// 	return <>{children}</>;
// };
