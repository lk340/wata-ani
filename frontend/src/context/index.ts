// ================= //
// ↓↓↓ Utilities ↓↓↓ //
// ================= //

import * as UserAgent from "@/context/utilities/user-agent";
import * as Location from "@/context/utilities/location";
import * as Windows from "@/context/utilities/windows";
import * as Scroll from "@/context/utilities/scroll";
import * as Theme from "@/context/utilities/theme";
import * as Mouse from "@/context/utilities/mouse";

// ============ //
// ↓↓↓ Data ↓↓↓ //
// ============ //

import * as AuthForm from "@/context/data/auth-form";
import * as Navbar from "@/context/data/navbar";
import * as Pagination from "@/context/data/pagination";
import * as NavbarOptionsCreate from "@/context/data/navbar-options-create";
import * as NavbarOptionsSettings from "@/context/data/navbar-options-settings";

// =============== //
// ↓↓↓ Exports ↓↓↓ //
// =============== //

export {
	UserAgent,
	Location,
	Windows,
	Scroll,
	Theme,
	Mouse,
	AuthForm,
	Navbar,
	Pagination,
	NavbarOptionsCreate,
	NavbarOptionsSettings,
};
