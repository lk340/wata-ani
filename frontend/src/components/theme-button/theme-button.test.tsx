import * as React from "react";
import * as Testing from "@testing-library/react";

import * as Context from "@/context";
import * as Tests from "@/utils/tests";

import { ThemeButton } from "./theme-button";

describe("ThemeButton Component", () => {
	it("matches the snapshot", () => {
		const { asFragment } = Tests.renderWithContext(<ThemeButton />);

		expect(asFragment(<ThemeButton />)).toMatchSnapshot();
	});

	it("toggles the theme when the user clicks on it", () => {
		const ThemeButtonWithMode = () => {
			const { theme } = Context.Theme.useThemeContext();

			return (
				<React.Fragment>
					<ThemeButton />
					<div data-testid="mode">{theme.state.mode}</div>
				</React.Fragment>
			);
		};

		const themeButtonWithMode = Tests.renderWithContext(<ThemeButtonWithMode />);

		const mode = themeButtonWithMode.getByTestId(/mode/);
		expect(mode).toHaveTextContent(/light/);

		const themeButton = themeButtonWithMode.getByTestId(/theme button component/);
		Testing.fireEvent.click(themeButton);
		expect(mode).toHaveTextContent(/dark/);
	});
});
