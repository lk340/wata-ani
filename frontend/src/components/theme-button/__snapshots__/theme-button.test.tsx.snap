// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeButton Component matches the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  margin: 0px 40px;
  cursor: pointer;
}

.c1 {
  width: 20px;
  height: 20px;
  position: absolute;
}

.c2 {
  width: 20px;
  height: 20px;
}

<div
    class="c0"
    data-testid="theme button component"
  >
    <img
      alt="navbar component light theme button"
      class="c1"
      src="[object Object]"
      style="opacity: 0;"
    />
    <img
      alt="navbar component dark theme button"
      class="c2"
      src="[object Object]"
    />
  </div>
</DocumentFragment>
`;
